mindmaps-graql/src/main/java/io/mindmaps/graql/internal/analytics/ShortestPathVertexProgram.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    @Override
+    public Set<String> getElementComputeKeys() {
+        return ELEMENT_COMPUTE_KEYS;
+    }
+
+    @Override
+    public Set<String> getMemoryComputeKeys() {
+        return MEMORY_COMPUTE_KEYS;
+    }
+
+    @Override
+    public Set<MessageScope> getMessageScopes(final Memory memory) {
+        if (memory.get(FOUND_DESTINATION)) return Collections.emptySet();
+        return messageScopeSet;
+    }
+
+    @Override
+    public void setup(final Memory memory) {
+        LOGGER.debug("ShortestPathVertexProgram Started !!!!!!!!");
+        memory.set(VOTE_TO_HALT, true);
+        memory.set(FOUND_DESTINATION, false);
+        memory.set(NEXT_DESTINATION, "");
+    }
+
+    @Override
+    public void safeExecute(final Vertex vertex, Messenger<String> messenger, final Memory memory) {
+        switch (memory.getIteration()) {
+            case 0:
+                if (selectedTypes.contains(Utility.getVertexType(vertex))) {
+                    String type = vertex.label();
+                    if (type.equals(Schema.BaseType.ENTITY.name()) || type.equals(Schema.BaseType.RESOURCE.name())) {
+                        // each role-player sends 1 to castings following incoming edges
+                        messenger.sendMessage(messageScopeIn, MESSAGE_FROM_ROLE_PLAYER);
+                    } else if (type.equals(Schema.BaseType.RELATION.name())) {
+                        // the assertion can also be a role-player, so sending 1 to castings following incoming edges
+                        messenger.sendMessage(messageScopeIn, MESSAGE_FROM_ROLE_PLAYER);
+                        // send -1 to castings following outgoing edges
+                        messenger.sendMessage(messageScopeOut, MESSAGE_FROM_ASSERTION);
+                    }
+                    // send message from the starting vertex
+                    String id = vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name());
+                    if (persistentProperties.get(START_ID).equals(id)) {
+                        LOGGER.debug("Found starting vertex");
+                        messenger.sendMessage(messageScopeIn, id);
+                        messenger.sendMessage(messageScopeOut, id);
+                    }
+                }
+                break;
+            case 1:
+                if (vertex.label().equals(Schema.BaseType.CASTING.name())) {
+//                    String id = vertex.value(Schema.ConceptProperty.ITEM_IDENTIFIER.name());
+                    Set<String> messageSet = new HashSet<>();
+                    boolean hasBothMessages = false;
+                    String fromVertex = null;
+                    Iterator<String> iterator = messenger.receiveMessages();
+                    while (iterator.hasNext()) {
+                        String message = iterator.next();
+                        if (message.length() == 1) messageSet.add(message);
+                        else fromVertex = message;
```

Comment:
is it a good idea to use the length of the vertex - could we just use more informative messages?